server/server/src/cold_chain/temperature_breach.rs

```diff
@@ -0,0 +1,188 @@
+    service_provider: Data<ServiceProvider>,
+    auth_data: Data<AuthData>,
+    breaches: web::Json<Vec<TemperatureBreach>>,
+) -> HttpResponse {
+    let store_id = match validate_request(request, &service_provider, &auth_data) {
+        Ok((_user, store_id)) => store_id,
+        Err(error) => {
+            let formatted_error = format!("{:#?}", error);
+            return HttpResponse::Unauthorized().body(formatted_error);
+        }
+    };
+    let breaches = breaches.into_inner();
+    if !validate_input(&breaches) {
+        return HttpResponse::BadRequest()
+            .body("Expecting a body with the array of temperature breaches");
+    };
+
+    let results = match upsert_temperature_breaches(service_provider, store_id, breaches).await {
+        Ok(response) => response,
+        Err(error) => return HttpResponse::InternalServerError().body(format!("{:#?}", error)),
+    };
+
+    HttpResponse::Ok()
+        .append_header(header::ContentType(mime::APPLICATION_JSON))
+        .json(&results)
+}
+
+fn validate_input(breaches: &Vec<TemperatureBreach>) -> bool {
+    breaches.iter().all(|breach| validate_breach(breach))
+}
+
+fn validate_breach(breach: &TemperatureBreach) -> bool {
+    match breach.end_unix_timestamp {
+        Some(end_unix_timestamp) => {
+            if end_unix_timestamp < 0 {
+                return false;
+            }
+        }
+        None => {}
+    }
+    if breach.start_unix_timestamp < 0 {
+        return false;
+    }
+    if breach.threshold_duration < 0 {
+        return false;
+    }
+    true
+}
+
+async fn upsert_temperature_breaches(
+    service_provider: Data<ServiceProvider>,
+    store_id: String,
+    breaches: Vec<TemperatureBreach>,
+) -> Result<Vec<Result<repository::TemperatureBreach, String>>, RepositoryError> {
+    let ctx = service_provider.context(store_id, SYSTEM_USER_ID.to_string())?;
+    let results = breaches
+        .into_iter()
+        .map(|breach| {
+            upsert_temperature_breach(&service_provider, &ctx, breach.clone()).map_err(|e| {
+                error!("{:#?} {:?}", e, breach);
+                e.to_string()
+            })
+        })
+        .collect();
+
+    Ok(results)
+}
+
+fn upsert_temperature_breach(
+    service_provider: &ServiceProvider,
+    ctx: &ServiceContext,
+    breach: TemperatureBreach,
+) -> anyhow::Result<repository::TemperatureBreach> {
+    let id = breach.id.clone();
+    let service = &service_provider.temperature_breach_service;
+    let sensor_service = &service_provider.sensor_service;
+    let sensor = sensor_service
+        .get_sensor(&ctx, breach.sensor_id.clone())
+        .map_err(|e| anyhow::anyhow!("Unable to get sensor {:?}", e))?;
+    let start_datetime = NaiveDateTime::from_timestamp_opt(breach.start_unix_timestamp, 0)
+        .context(format!(
+            "Unable to parse timestamp {}",
+            breach.start_unix_timestamp
+        ))?;
+
+    let duration = match breach.end_unix_timestamp {
```

Comment:
Do we need to set duration if end_time doesn't exist (to use now as end date time ?), i see in mSupply code we don't actually set duration in cold chain api (so it must be calcuated on the go when we display breaches), berlinger integration does add breah duration, so I think it's a good idea to also set it here.

Btw I need to make an issue to add unit to the duration fields so that schema make it clear.

Reply:
My thinking was that if the end_time doesn't exist then we can't know duration. Setting the duration to now-start is only correct now. Any time in the future, that duration value will be wrong, so it's better to set it to 0 and update it when the breach finishes

Reply:
From telegram chat is sounds like duration should be shown, which we can do even if duration is set to 0, but I think it's easier to set it now (but it would have to be relative to the latest log ?). Also to sort the column we would have to set it in db rather then deducing during presentation